resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java

```diff
@@ -0,0 +1,289 @@
+            buffered = bufIn;
+        }
+
+        final Request request = client.newRequest(invocation.getUri());
+        final CompletableFuture<T> future = new RequestFuture<T>(request);
+        // readEntity calls releaseConnection which calls cancel, so don't let that interrupt us
+        final AtomicBoolean completing = new AtomicBoolean();
+
+        invocation.getMutableProperties().forEach(request::attribute);
+        request.method(invocation.getMethod());
+        invocation.getHeaders().asMap().forEach((h, vs) -> vs.forEach(v -> request.header(h, v)));
+        configureTimeout(request);
+        if (request.getAttributes().get(FOLLOW_REDIRECTS) == Boolean.FALSE) {
+            request.followRedirects(false);
+        }
+
+        if (invocation.getEntity() != null) {
+            final ByteArrayOutputStream os = new ByteArrayOutputStream();
+            try {
+                invocation.writeRequestBody(os);
+            } catch (IOException e) {
+                future.completeExceptionally(e);
+                if (callback != null) {
+                    callback.failed(e);
+                }
+                return future;
+            }
+            request.content(new BytesContentProvider(os.toByteArray()));
+        }
+
+        request.send(new Response.Listener.Adapter() {
+            private ClientResponse cr;
+            private JettyResponseStream stream = new JettyResponseStream();
+
+            @Override
+            public void onHeaders(Response response) {
+                cr = new JettyClientResponse(invocation.getClientConfiguration(), stream, () -> {
+                    if (!completing.get()) {
+                        future.cancel(true);
+                    }
+                });
+                cr.setProperties(invocation.getMutableProperties());
+                cr.setStatus(response.getStatus());
+                cr.setHeaders(extract(response.getHeaders()));
+                if (!buffered) {
+                    complete();
+                }
+            }
+
+            @Override
+            public void onContent(Response response, ByteBuffer buf) {
+                final ByteBufferPool bufs = client.getByteBufferPool();
+                final ByteBuffer copy = bufs.acquire(buf.remaining(), false);
+                copy.limit(buf.remaining());
+                copy.put(buf);
+                copy.flip();
+                stream.offer(copy, new ReleaseCallback(bufs, copy));
+            }
+
+            @Override
+            public void onSuccess(Response response) {
+                if (buffered) {
+                    try {
+                        complete();
+                    } catch (Exception e) {
+                        onFailure(response, e);
+                    }
+                }
+            }
+
+            @SuppressWarnings("unchecked")
+            private void complete() {
+                completing.set(true);
+                // TODO: dangerous cast, see javadoc!
```

Comment:
What about plans for this todo?

Reply:
This might not actually be a TODO, you can only get an invalid cast here if the caller "uses" the class wrong.  But it might be nicer to throw an explicit error with a better message about what went wrong.